Core/Spells: fix Priest spell Penance and rewrite Power Word:Shield

HannibalRoG opened the pull request:
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  Correct implement of talent Penitent for Spell Penance and re-write Power Word Shield.


**Target branch(es):** 3.3.5/master

- master

**Issues addressed:** Closes #  (insert issue tracker number)
Closes #18520

**Tests performed:** (Does it build, tested in-game, etc.)
Compiled and tested in game.

HannibalRoG committed:
Fix for Spell Priest Penance
Closes #18520

HannibalRoG committed:
Update spell_priest.cpp

lorac2k14 commented:
whats going on with your spacing it is all over the place? make sure you are following [C++ Development Standards](https://trinitycore.atlassian.net/wiki/pages/viewpage.action?pageId=2130103)

Keader commented:
use 4 spaces instead tabs

HannibalRoG committed:
Fix tab with spaces.

HannibalRoG committed:
Merge branch 'Spell_Priest_Penance' of https://github.com/Edreisantafe/TrinityCore into Spell_Priest_Penance

# Conflicts:
#	src/server/scripts/Spells/spell_priest.cpp

HannibalRoG committed:
Try to fix Travis.

HannibalRoG committed:
Delete variable non used.

HannibalRoG committed:
Changes in code, sorry.

Commit:
Update and rename RoG_2017_01_12_04_world.sql to 2017_02_28_19033_world.sql

- using TC file naming standard for PR world DB update SQL
- minor spelling correction in comment text
- add missing NewLine (CR/LF) at end of file
- remove empty first line

tkrokli commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+        void HandleOnRemove(const AuraEffect* aurEff, AuraEffectHandleModes mode)
+        {
+            Player* caster = GetCaster()->ToPlayer();
+            int32 max = aurEff->GetBase()->GetMaxDuration();
```
```
src/server/scripts/Spells/spell_priest.cpp:1254:19: fatal error: 
      unused variable 'max' [-Wunused-variable]
            int32 max = aurEff->GetBase()->GetMaxDuration();
                  ^
```

HannibalRoG committed:
Yes, travis, i am learning.

P-Kito commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+        SPELL_PRIEST_RENEWED_HOPE_EFFECT        = 197470,
+        SPELL_PRIEST_VOID_SHIELD                = 199144,
+        SPELL_PRIEST_VOID_SHIELD_EFFECT         = 199145
+    };
```
missing line

P-Kito commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+            Player* caster = GetCaster()->ToPlayer();
+
+            if (caster->HasAura(SPELL_PRIEST_BODY_AND_SOUL))
+                caster->CastSpell(caster, SPELL_PRIEST_BODY_AND_SOUL_ACTIVE,true);
```
Codestyle ", true"

P-Kito commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+    };
+    class spell_pri_power_word_shield_SpellScript : public SpellScript
+    {
+
```
remove this line :P

P-Kito commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+            amount = (int32)((sp * 5.5f) + mastery) * (1 + versadmg);
+        }
+        
+        void Absorb(AuraEffect* aurEff, DamageInfo & dmgInfo, uint32 & absorbAmount)
```
this function is not crash safe, make sure GetEffect is checked and ToPlayer is checked, also rename "n" and "m" to something easier to understand. 
Another suggestion:

```c++
if (n <= m)
    absorbAmount += (int32)dmgInfo.GetDamage();
```

P-Kito commented:
Fix the codestyle, there are alot of unclear variable names, missing spaces or too much space.

I am not going to comment them all, but please double check the code and fix the codestyle.
(I already commented some)

P-Kito commented on sql/updates/world/master/2017_02_28_19033_world.sql:
```diff
@@ -0,0 +1,5 @@
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE ScriptName = 'spell_pri_power_word_shield';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES ('17', 'spell_pri_power_word_shield');
```
Missing new line at the end :-)

Aokromes replied:
nope, no "no newline" mark on diff.

P-Kito replied:
My bad, didn't look like it :P

tkrokli replied:
I fixed it in my commit 98bb86c162d354a512725945f37144921a22befc

offl replied:
Never use apostrophes for numbers.
```sql
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (17, 'spell_pri_power_word_shield');
```

Aokromes commented:
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Spells/spell_priest.cpp:1254:19: fatal error: 
      unused variable 'max' [-Wunused-variable]
            int32 max = aurEff->GetBase()->GetMaxDuration();
                  ^

HannibalRoG committed:
Merge branch 'Spell_Priest_Penance' of https://github.com/Edreisantafe/TrinityCore into Spell_Priest_Penance

Edreisantafe commented:
I am taking note, lol.

HannibalRoG committed:
Fix the notes from Trinity Team.

HannibalRoG committed:
This reverts commit 855d9ab4d1af8c56220003bd1a963fa3f7177ba2.

HannibalRoG committed:
Comments travis, comments.

Edreisantafe commented:
Well, at last it seems Travis is happy.

Edreisantafe commented:
Well, at last it seems Travis is happy.

Edreisantafe commented:
Compiled and tested, I think it's Blizzlike. Any corrections?

tkrokli commented:
Code style: add 1 CR/LF at the end of file (EOF) of `spell_priest.cpp` - your typing cursor should always be able to move 1 line below the last line of code in any of the files.

offl commented on sql/updates/world/master/2017_02_28_19033_world.sql:
```diff
@@ -0,0 +1,5 @@
+-- Class: Priest
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE ScriptName = 'spell_pri_power_word_shield';
```
Always use the grave accent for table or column names.
```sql
DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_pri_power_word_shield';
```

offl commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -755,6 +757,7 @@ class spell_pri_pain_and_suffering_proc : public SpellScriptLoader
                         aur->SetDamage(caster->SpellDamageBonusDone(target, aur->GetSpellInfo(), damage, DOT, aur->GetSpellEffectInfo()) * aur->GetDonePct());
                         aur->CalculatePeriodic(caster, false, false);
                         aur->GetBase()->RefreshDuration();
+						
```
Excess line.

offl commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1269,4 +1321,4 @@ void AddSC_priest_spell_scripts()
     new spell_pri_vampiric_embrace_target();
     new spell_pri_vampiric_touch();
-}
+}
```
"No newline at end of file"

HannibalRoG committed:
More corrections.

offl commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1241,6 +1181,117 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
+        }
+    };
+
+    class spell_pri_power_word_shield_AuraScript :public AuraScript
```
Missing space.
```cpp
    class spell_pri_power_word_shield_AuraScript : public AuraScript
```

offl commented on sql/updates/world/master/2017_02_28_19033_world.sql:
```diff
@@ -0,0 +1,5 @@
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_pri_power_word_shield';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES ('17', 'spell_pri_power_word_shield');
```
Again.
```sql
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (17, 'spell_pri_power_word_shield');
```

HannibalRoG committed:
More...

HannibalRoG committed:
...

wonder2011 commented:
I test this code and is working fine.

Edreisantafe commented:
Doing a research with my starter edition account I found two problems that are not contemplated in the script, one is that the amount of absorption of the shield is not so high, that is, the formula seems not correct, and two, the aura Updates every time the shield absorbs damage, this is more of a visual problem.
I will review the code and add these two problems, I do not know if they should wait for them to be fixed to include this or add these first and then the arrangements.
If you could give me some suggestion as to the formula I would be grateful.

Edreisantafe commented:
Bueno, más investigaciones han arrojado lo siguiente:
La fórmula parece estar bastante cercana a la realidad, comparando las imágenes retail 01.jpg y trinity 01.jpg vemos poca diferencia, de hecho en el script está por debajo.
![retail 01](https://cloud.githubusercontent.com/assets/19369492/22774760/8ce39b10-ee76-11e6-9d07-72728462413b.jpg)
![trinity 01](https://cloud.githubusercontent.com/assets/19369492/22774763/921125e4-ee76-11e6-94be-bf99c3aa18d5.jpg)
Además con un objeto con 2 de versatilidad los números son bastante similares.
![retail 02](https://cloud.githubusercontent.com/assets/19369492/22774773/9bf39dd0-ee76-11e6-998c-f467c00d6f5d.jpg)
![trinity 02](https://cloud.githubusercontent.com/assets/19369492/22774774/9e53a3fe-ee76-11e6-9ebc-6f64c9e74dca.jpg)
Lo que sucede es que cuando tenemos, por ejemplo, 8 de versatilidad el número aumenta drásticamente 771 puntos, este objeto "Leaving Root" no debiera tener tanta versatilidad en nivel 20.
Las pruebas iniciales me habían mostrado grandes números de absorción debido a la cantidad de versatilidad que presentaba el PJ.
Sólo queda implementar que se actualice el aura con el número restante de absorción de daño.
Nótece el bug de regeneración de maná, en el retail es de 80 puntos y en trinity 200 cada 5s, esto hace que el maná del sacerdote sea infinito.

HannibalRoG commented:
Any news with this PR?

tkrokli commented:
If there are no objections or mentions of things to change, we could merge it before Monday.

Commit:
Update spell_priest.cpp

Minor cosmetic changes in the spell name comments.

Commit:
Update and rename 2017_02_28_19033_world.sql to 2017_02_13_00_world.sql

- condensed comment lines from 3 to 2
- renamed file to be correct after merge

Commit:
Update spell_priest.cpp

- keep the the enum list in alphabetical order

tkrokli commented:
Thank you for the PR, your work is appreciated. 👍

wonder2011 commented:
The PVP Talents are implemented ?
